Add the ability to download pods in parallel

sethfri opened the pull request:
Currently, when CocoaPods installs a pod, it first downloads the pod if it's not local or already cached, and then it does the installation for the pod. It does this combined download/installation step serially for each pod that needs to be installed.

While parallelizing the pod installation process itself might be a larger effort given the potential assumptions built in throughout the codebase that installation is performed serially, we can still improve `pod install` times significantly by parallelizing the downloads.

To do this, we first need to separate the download step from the installation step, instead of doing downloading and installing at the same time for each pod. Using the opt-in `:parallel_pod_downloads` option, CocoaPods now downloads all pods in parallel, and then moves on to serially install each pod. By default, this is done with 40 threads, which is the sweet spot I found in my testing. However, the number of threads is configurable via the `:parallel_pod_download_thread_pool_size` option.

The PR is likely easiest to review commit-by-commit:
* Add `parallel_pod_downloads` option
* Extract `PodSourceDownloader` out of `PodSourceInstaller`
* Implement parallel downloads
* Add `parallel_pod_download_thread_pool_size` option
* Extract `section_title` method

In my testing of a proprietary project on a 2020 M1 MacBook Pro, I got the following results for a completely clean `pod install` (after running `rm -rf Pods && rm -rf ~/Library/Caches/CocoaPods`):

```
No parallelization
pod install  82.63s user 33.51s system 25% cpu 7:35.21 total

10 threads
pod install  93.61s user 38.64s system 78% cpu 2:47.81 total

20 threads
pod install  94.86s user 40.38s system 89% cpu 2:31.75 total

30 threads
pod install  92.38s user 39.68s system 92% cpu 2:22.82 total

40 threads
pod install  94.31s user 43.23s system 99% cpu 2:18.13 total

50 threads
pod install  97.80s user 42.16s system 68% cpu 2:28.04 total
```

This means that using 40 threads to parallelize downloads instead of no parallelization results in a 70% speedup for clean pod install times, saving 5 min 17 sec for this particular project.

sethfri commented:
cc @dnkoutso @amorde

dnkoutso commented:
cc @igor-makarov

sethfri commented on lib/cocoapods/installer.rb:
```diff
@@ -502,26 +504,37 @@ def verify_no_lockfile_changes!
+
+      sorted_root_specs = root_specs.sort_by(&:name)
+
+      # Download pods in parallel before installing if the option is set
```
That's true, but I wonder if it would be safer to leave the parallel option in place to act as a feature flag of sorts for the new downloading logic, just in case something goes wrong

dnkoutso replied:
It is true that semantically we will introduce another thread even if its just 1 versus the current implementation. I am OK to proceed as such.

amorde commented on lib/cocoapods/installer/pod_source_installer.rb:
```diff
@@ -186,6 +153,13 @@ def root
+      # @return [Boolean] whether the source has already been downloaded prior
+      #         to the installation process.
+      #
+      def downloaded?
```
are the methods here and below still used in this file now that they've been added to the new class `PodSourceDownloader`?

sethfri replied:
Yes, it's still used in the `download_source` method, which calls into the `PodSourceDownloader`

dnkoutso commented:
@sethfri do we need to split any specs (tests) as well?

sethfri commented:
> @sethfri do we need to split any specs (tests) as well?

Absolutely. I haven't touched the tests yet because I wanted to verify that the approach was good first

sethfri commented on lib/cocoapods/installer/installation_options.rb:
```diff
@@ -190,6 +190,17 @@ def hash
+      #
+      # Default: 40
+      #
+      option(:parallel_pod_download_thread_pool_size, 40, :boolean => false)
```
Do you have any recs on a particular ratio? Not sure of the best way to calculate

dnkoutso replied:
I wonder if it isnt much about CPU but rather network bandwidth which is harder to gauge so maybe CPU cores is good enough?

dnkoutso replied:
I guess I wouldnt worry much about it now imo

amorde replied:
you could start with a 1:1 ratio. but yeah I suppose this is ok for an initial implementation and can be optimized further later down the road

sethfri commented on lib/cocoapods/installer.rb:
```diff
@@ -576,8 +622,20 @@ def install_source_of_pod(pod_name)
     end
 
-    # Cleans the sources of the Pods if the config instructs to do so.
+    # Install the Pods. If the resolver indicated that a Pod should be
+    # installed and it exits, it is removed and then reinstalled. In any case if
+    # the Pod doesn't exits it is installed.
```
Good catch, fixed

dnkoutso commented on lib/cocoapods/installer.rb:
```diff
@@ -1,4 +1,5 @@
 require 'active_support/core_ext/string/inflections'
+require 'concurrent/executor/fixed_thread_pool'
```
we can move this `require` statement only if the installation option is enabled.

sethfri replied:
Nice, didn't know you could do that. Changed

sethfri commented:
@dnkoutso @amorde I split out some of the tests from the `PodSourceInstaller` into `PodSourceDownloader`, and also added some new tests. Should be good for a proper review now

lockieluke commented:
how do i utilise this feature

chenxGen commented:
> how do i utilise this feature

Add the following lines to your `Podfile`.

```ruby
install! 'cocoapods',
         :parallel_pod_downloads => true,
         :parallel_pod_download_thread_pool_size => 40
```

FrederickEngelhardt commented:
Thanks for doing this!
